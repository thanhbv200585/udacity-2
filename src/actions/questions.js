import { saveQuestionAnswer, saveQuestion } from "../utils/api";
import {
  addQuestionToAuthor,
  addAnsweredQuestionToUser,
} from "../actions/users";
import { showLoading, hideLoading } from "react-redux-loading-bar";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ANSWER_QUESTION = "ANSWER_QUESTION";
export const ADD_QUESTION = "ADD_QUESTION";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

function answerQuestion({ authedUser, qid, answer }) {
  return {
    type: ANSWER_QUESTION,
    authedUser,
    qid,
    answer,
  };
}

export function handleAnswerQuestion(info) {
  return (dispatch) => {
    dispatch(showLoading());
    dispatch(answerQuestion(info));
    dispatch(addAnsweredQuestionToUser(info));
    dispatch(hideLoading());

    return saveQuestionAnswer(info).catch((e) => {
      console.warn("Error in saveQuestionAnswer: ", e);
      alert("There was an error. Try again");
    });
  };
}

function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function handleAddQuestion(optionOne, optionTwo) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    return saveQuestion({
      optionOneText: optionOne,
      optionTwoText: optionTwo,
      author: authedUser,
    }).then((question) => {
      dispatch(showLoading());
      dispatch(addQuestion(question));
      dispatch(addQuestionToAuthor(question));
      dispatch(hideLoading());
    })
    .catch((e) => {
      console.warn("Error in saveQuestion: ", e);
      alert("There was an error. Try again");
    });
  };
}
