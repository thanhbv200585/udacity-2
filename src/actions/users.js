export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_ANSWERED_QUESTION = "ADD_ANSWERED_QUESTION";
export const ADD_QUESTION_TO_AUTHOR = "ADD_QUESTION_TO_AUTHOR";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function addAnsweredQuestionToUser({ qid, authedUser, answer }) {
  return {
    type: ADD_ANSWERED_QUESTION,
    qid,
    authedUser,
    answer,
  };
}

export function addQuestionToAuthor(question) {
  return {
    type: ADD_QUESTION_TO_AUTHOR,
    question,
  };
}
