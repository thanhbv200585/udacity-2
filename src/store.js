import reducer from "./reducers";
import { createStore } from "redux";
import middleware from "./middleware";

const store = createStore(reducer, middleware);

export { store };
