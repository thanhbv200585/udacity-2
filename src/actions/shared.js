import { getInitialData } from "../utils/api";
import { receiveQuestions } from "./questions";
import { receiveUsers } from "./users";
import { setAuthedUser } from "./authedUser";
import { showLoading, hideLoading } from "react-redux-loading-bar";

const AUTHED_ID = null

export function handleInitialData() {
  return (dispatch) => {
    dispatch(showLoading());
    return getInitialData().then(({ users, questions }) => {
      dispatch(receiveQuestions(questions));
      dispatch(receiveUsers(users));
      dispatch(setAuthedUser(AUTHED_ID));
      dispatch(hideLoading());
    });
  };
}
