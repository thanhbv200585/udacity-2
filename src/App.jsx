import { Route, Routes } from "react-router-dom";
import { useEffect } from "react";
import Dashboard from "./pages/Dashboard";
import { connect } from "react-redux";
import { handleInitialData } from "./actions/shared";
import LoadingBar from "react-redux-loading-bar";
import NewQeustion from "./pages/NewQuestion";
import QuestionPage from "./pages/QuestionPage";
import LoginPage from "./pages/LoginPage";
import LeaderBoard from "./pages/LeaderBoard";
import NotFound from "./pages/NotFound";
import ProtectedRoutes from "./components/layout/ProtectedRoutes";

function App(props) {
  useEffect(() => {
    props.dispatch(handleInitialData());
  });

  return (
    <>
      <LoadingBar />
      {props.loading === true ? null : (
        <Routes>
          <Route element={<ProtectedRoutes />}>
            <Route path="/" exact element={<Dashboard />} />
            <Route path="/add" element={<NewQeustion />} />
            <Route path="/questions/:id" element={<QuestionPage />} />
            <Route path="/leaderboard" element={<LeaderBoard />} />
          </Route>
          <Route path="/login" element={<LoginPage />} />
          <Route path="*" element={<NotFound />} />
        </Routes>
      )}
    </>
  );
}

const mapStateToProps = ({ users }) => ({
  loading: users === null,
});

export default connect(mapStateToProps)(App);
