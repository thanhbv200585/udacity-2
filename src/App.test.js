import { render, screen } from "@testing-library/react";
import App from "./App";
import { store } from "./store";
import "@testing-library/jest-dom";
import { BrowserRouter as Router } from "react-router-dom";
import { Provider } from "react-redux";
import { setAuthedUser } from "./actions/authedUser";

describe("App", () => {
  it("is expected to render the component", () => {
    const view = render(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    );
    expect(view).toMatchSnapshot();
  });
  it("It is expected to load the Login Page if no authed User is logged.", () => {
    render(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    );
    expect(screen.getByRole("button", { name: "Login" })).toBeInTheDocument();
  });

  it("It is expected to give access to the app if a user is logged in.", async () => {
    const user = "mtsamis";
    await store.dispatch(setAuthedUser(user));
    render(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    );
    expect(
      screen.queryByRole("button", { name: "Login" })
    ).not.toBeInTheDocument();
  });
});
