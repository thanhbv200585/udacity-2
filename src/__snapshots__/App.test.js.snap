// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App is expected to render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="direction: ltr;"
      >
        <div
          class=""
          style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
        />
        <div
          style="display: table; clear: both;"
        />
      </div>
      <div
        class="flex items-center justify-center p-3 h-[100vh] bg-slate-600"
      >
        <form
          class="max-w-md w-full p-4 bg-white rounded-lg flex flex-col gap-4"
        >
          <h1
            class="text-center uppercase text-2xl font-bold"
          >
            Login
          </h1>
          <input
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 focus-within:outline-none"
            id="username"
            name="username"
            placeholder="Enter username"
            type="text"
            value="sarahedo"
          />
          <input
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 focus-within:outline-none"
            id="password"
            name="password"
            placeholder="Enter password"
            type="Password"
            value="password123"
          />
          <button
            class="text-white focus:outline-none w-full focus:ring-4 font-medium rounded-lg text-sm px-5 py-2.5 disabled:bg-gray-500 bg-blue-700 hover:bg-blue-800 focus:ring-blue-300"
            type="submit"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="direction: ltr;"
    >
      <div
        class=""
        style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
      />
      <div
        style="display: table; clear: both;"
      />
    </div>
    <div
      class="flex items-center justify-center p-3 h-[100vh] bg-slate-600"
    >
      <form
        class="max-w-md w-full p-4 bg-white rounded-lg flex flex-col gap-4"
      >
        <h1
          class="text-center uppercase text-2xl font-bold"
        >
          Login
        </h1>
        <input
          class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 focus-within:outline-none"
          id="username"
          name="username"
          placeholder="Enter username"
          type="text"
          value="sarahedo"
        />
        <input
          class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 focus-within:outline-none"
          id="password"
          name="password"
          placeholder="Enter password"
          type="Password"
          value="password123"
        />
        <button
          class="text-white focus:outline-none w-full focus:ring-4 font-medium rounded-lg text-sm px-5 py-2.5 disabled:bg-gray-500 bg-blue-700 hover:bg-blue-800 focus:ring-blue-300"
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
